vicodyn/src/vicodyn/proxy.cpp

```diff
@@ -108,26 +108,97 @@ class safe_stream_t {
     }
 };
 
-class vicodyn_dispatch_t : public std::enable_shared_from_this<vicodyn_dispatch_t> {
-    using protocol = io::protocol<app_tag>::scope;
-    using clock_t = peers_t::clock_t;
+class buffer_t {
+    std::string event_;
+    hpack::headers_t headers_;
 
-    struct error_buffer_t {
-        hpack::headers_t headers;
-        std::error_code ec;
-        std::string msg;
-    };
+    std::vector<std::string> chunks_;
+    std::vector<hpack::headers_t> chunk_headers_;
+    boost::optional<hpack::headers_t> choke_;
+    bool failed_ = false;
 
-    struct buffer_t {
-        std::string event;
-        hpack::headers_t headers;
+    std::size_t size_ = 0;
+    std::size_t max_size_;
 
-        std::vector<std::string> chunks;
-        std::vector<hpack::headers_t> chunk_headers;
 
-        boost::optional<error_buffer_t> error;
-        boost::optional<hpack::headers_t> choke;
-    };
+private:
+    auto account_headers(const hpack::headers_t& headers) -> bool {
+        if (overfull()) {
+            return false;
+        }
+        for (const auto& header : headers) {
+            size_ += header.http2_size();
+        }
+        return !overfull();
+    }
+
+    auto account_data(const std::string& data) -> bool {
+        if (overfull()) {
+            return false;
+        }
+        size_ += data.size();
+        return !overfull();
+    }
+
+
+public:
+    buffer_t(std::size_t max_size)
+        : max_size_(max_size) {
+    }
+
+    auto overfull() const -> bool {
+        return size_ > max_size_;
+    }
+
+    auto failed() const -> bool {
+        return failed_;
+    }
+
+    auto set_event(const hpack::headers_t& headers, const std::string& event) -> void {
+        if (account_headers(headers) && account_data(event)) {
+            headers_ = headers;
+            event_ = event;
+        }
+    }
+
+    auto event() const -> const std::string& {
+        return event_;
+    }
+
+    auto headers() const -> const hpack::headers_t& {
+        return headers_;
+    }
+
+    auto chunk(const hpack::headers_t& headers, const std::string& data) -> void {
+        if (account_headers(headers) && account_data(data)) {
+            chunk_headers_.push_back(headers);
+            chunks_.push_back(data);
+        }
+    }
+
+    auto close(const hpack::headers_t& headers) -> void {
+        if (account_headers(headers)) {
+            choke_ = headers;
+        }
+    }
+
+    auto error(const hpack::headers_t& /*headers*/, const std::error_code& /*ec*/, const std::string& /*msg*/) -> void {
+        failed_ = true;
+    }
+
+    auto flush(safe_stream_t& stream) const -> void {
```

Comment:
Could flush be called several times for same chunk sequence? Should both chunk and headers vectors be cleared after flush done?